File: apps/web/src/features/agent/chat-items.ts
import type { ThreadMessage, ToolCallMessage } from "@ddl/core";

export type ChatItem =
  | { kind: "message"; message: ThreadMessage }
  | { kind: "tools"; id: string; calls: readonly ToolCallMessage[] };

const finished = (message: ThreadMessage): message is ToolCallMessage =>
  message.kind === "tool_call" && message.status === "ok";

/**
 * The chat's rows: consecutive tool calls that finished fine collapse into one group (two or
 * more); running and failed calls stay visible, and so does a call that just finished while it's
 * the thread's last message, so its ✓ shows before the next step folds it in.
 */
export function chatItems(messages: readonly ThreadMessage[]): ChatItem[] {
  const items: ChatItem[] = [];
  let i = 0;
  while (i < messages.length) {
    const message = messages[i]!;
    if (!finished(message)) {
      items.push({ kind: "message", message });
      i++;
      continue;
    }
    let end = i;
    while (end < messages.length && finished(messages[end]!)) end++;
    const groupEnd = end === messages.length ? end - 1 : end;
    if (groupEnd - i >= 2) {
      items.push({
        kind: "tools",
        id: message.id,
        calls: messages.slice(i, groupEnd) as ToolCallMessage[],
      });
    } else {
      for (let j = i; j < groupEnd; j++) items.push({ kind: "message", message: messages[j]! });
    }
    for (let j = groupEnd; j < end; j++) items.push({ kind: "message", message: messages[j]! });
    i = end;
  }
  return items;
}

File: apps/macos/Packages/DailyDoListAgent/Sources/DailyDoListAgent/Chat/ChatItems.swift
import DailyDoListModels

/// One row of the chat: a message, or tool calls. Consecutive tool calls that succeeded collapse
/// into one run ("Used 6 tools"); a running call, a failure and a blocked call each stay a row of
/// their own.
struct ChatItem: Identifiable, Sendable {
  enum Content: Sendable {
    case message(ThreadMessage)
    /// One tool call, or several finished ones in a row.
    case tools([ToolCallMessage])
  }

  /// The row's identity: the message's id (the optimistic one it replaced, for the user's own
  /// messages), or the first call's (a run keeps its id as it grows).
  let id: String
  let content: Content

  /// - Parameter aliases: daemon message ids → the ids their rows had (`optimisticReplacements`).
  static func make(_ messages: [ThreadMessage], aliases: [String: String] = [:]) -> [ChatItem] {
    var items: [ChatItem] = []
    var run: [ToolCallMessage] = []
    func flush() {
      if let first = run.first { items.append(ChatItem(id: first.id, content: .tools(run))) }
      run = []
    }
    for message in messages {
      guard case .toolCall(let call) = message else {
        flush()
        items.append(ChatItem(id: aliases[message.id] ?? message.id, content: .message(message)))
        continue
      }
      if call.status == .ok {
        run.append(call)
      } else {
        flush()
        items.append(ChatItem(id: call.id, content: .tools([call])))
      }
    }
    flush()
    return items
  }
}
